content/fewer-braces.md

```diff
@@ -0,0 +1,162 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: ready for review
+status: implemented
+title: SIP-NN Fewer Braces
+---
+
+**By: Martin Odersky**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| July 1st 2022 | Initial Draft      |
+
+## Summary
+
+The current state of Scala 3 makes braces optional around blocks and template definitions (i.e. bodies of classes, objects, traits, enums, or givens). This SIP proposes to allow optional braces also for function arguments.
+The advantages of doing so is that the language feels more systematic, and programs become typographically cleaner.
+The changes have been implemented and and made available under the language import `language.experimental.fewerBraces`. The proposal here is to make them available without a language import instead.
+
+
+## Motivation
+
+After extensive experience with the current indentation rules I conclude that they are overall a big success.
+However. they still feel incomplete and a bit unsystematic since we can replace `{...}` in the majority of situations, but there are also important classes of situations where braces remain mandatory. In particular, braces are currently needed around blocks as function arguments.
+
+It seems very natural to generalize the current class syntax indentation syntax to function arguments. In both cases, an indentation block is started by a colon at the end of a line.
+
+
+## Proposed solution
+
+The proposed solution is described in detail in https://dotty.epfl.ch/docs/reference/other-new-features/indentation.html#variant-indentation-marker--for-arguments. I inline the relevant sections here:
+
+First, here is the spec for colons at ends of lines for template bodies. This is part of official Scala 3. I cited it here for context.
+
+> A template body can alternatively consist of a colon followed by one or more indented statements. To this purpose we introduce a new `<colon>` token that reads as
+the standard colon "`:`" but is generated instead of it where `<colon>`
+is legal according to the context free syntax, but only if the previous token
+is an alphanumeric identifier, a backticked identifier, or one of the tokens `this`, `super`, "`)`", and "`]`".
+
+> An indentation region can start after a `<colon>`. A template body may be either enclosed in braces, or it may start with
+`<colon> <indent>` and end with `<outdent>`.
+Analogous rules apply for enum bodies, type refinements, and local packages containing nested definitions.
+
+Generally, the possible indentation regions coincide with those regions where braces `{...}` are also legal, no matter whether the braces enclose an expression or a set of definitions. There is so far one exception, though: Arguments to functions can be enclosed in braces but they cannot be simply indented instead. Making indentation always significant for function arguments would be too restrictive and fragile.
```

Comment:
Do you mind elaborating on the drawbacks of the status quo? Why is this an important problem?

Also, the brace-less syntax can be used with function arguments:

~~~ scala
xs.map(x =>
  val y = x - 1
  y * y
)
~~~

This was not possible in Scala 2.

Reply:
Yes, but that does not work in all cases, which is why I would recommend to stick to braces. E.g. 
```scala
times(10)(
  val x = ...
  x * x
)
```
does not work.